osd: add cls_cxx_map_remove_range()

cbodley opened the pull request:
this adds a new `CEPH_OSD_OP_OMAPRMKEYRANGE` for the ObjectStore's `OP_OMAP_RMKEYRANGE`, and exposes it in the cls api as `cls_cxx_map_remove_range()` and the librados api as `rados_write_op_omap_rm_range()` and `ObjectWriteOperation::omap_rm_range()`

this is used by cls trim operations to avoid sending separate osd ops for each key with `cls_cxx_map_remove_key()`

Fixes: http://tracker.ceph.com/issues/19975

amitkumar50 commented on src/cls/log/cls_log.cc:
```diff
@@ -242,44 +242,25 @@ static int cls_log_trim(cls_method_context_t hctx, bufferlist *in, bufferlist *o
 
-#define MAX_TRIM_ENTRIES 1000
-  size_t max_entries = MAX_TRIM_ENTRIES;
+  CLS_LOG(20, "removing keys in range [%s, %s)",
```
@cbodley Extra spaces to be removed.

amitkumar50 commented on src/test/cls_log/test_cls_log.cc:
```diff
@@ -326,18 +297,69 @@ TEST(cls_rgw, test_log_trim)
-    bufferlist obl;
-    ASSERT_EQ(0, ioctx.operate(oid, rop, &obl));
+    log1.assign(std::make_move_iterator(entries.begin()),
+                std::make_move_iterator(entries.end()));
```
@cbodley tab indent

tchaikov commented:
@jdurgin ping?

cbodley commented:
rebased. `ceph_test_cls_rgw` is still passing, but `ceph_test_cls_log` depends on https://github.com/ceph/ceph/pull/18610 to pass `cls_log.test_log_add_same_time` and `cls_log.trim_by_marker`

cbodley commented:
rebased after https://github.com/ceph/ceph/pull/18610 merged, unit tests passing again

cbodley commented:
hi @jdurgin, i would love your feedback on this when you have the time. log trimming isn't a bottleneck for rgw, but it's currently way more expensive for the osd than it should be

@mattbenjamin mentioned that there had been some issues with range drops in rocksdb that might be relevant to this. if it's still a concern, could you (or @aclamk / @rzarzynski) expand on that?

tchaikov commented:
@jdurgin @neha-ojha ping?

neha-ojha commented:
needs rebase

cbodley commented:
rebased and updated OMAPRMKEYRANGE to the next available value 43

jdurgin commented on src/cls/log/cls_log.cc:
```diff
@@ -230,44 +230,25 @@ static int cls_log_trim(cls_method_context_t hctx, bufferlist *in, bufferlist *o
 
-  int rc = cls_cxx_map_get_vals(hctx, from_index, log_index_prefix, max_entries, &keys, &more);
-  if (rc < 0)
+  int rc = cls_cxx_map_remove_range(hctx, from_index, to_index);
```
the only thing that worries me about this is the lack of bound on number of entries to be removed in a single transaction. does rocksdb implement this in a way that is constant time, inserting implicit tombstones across the whole range, or is it still linear in the number of keys in the range?

cbodley replied:
good question! this was based on the naive assumption that the database would do this more efficiently in bulk than as separate keys/ranges. now that i look more closely at `RocksDBStore::RocksDBTransactionImpl::rm_range_keys()`, i see that by default (rocksdb_enable_rmrange=false) it will issue a separate rocksdb::WriteBatch::Delete() call for each key. the DeleteRange() path with rocksdb_enable_rmrange=true looks promising, but there appear to be performance issues related to its range tombstones (see https://github.com/facebook/rocksdb/issues/3959). i see that @yuyuyu101 was involved in this part of RocksDBStore, maybe you can offer some insight here?

regardless, i'm happy to rework this in a way that's closer to the current strategy, where it lists keys first in order to limit the range we pass to cls_cxx_map_remove_range(), then loops until ENODATA in the client - that would still be a huge improvement over the separate calls to cls_cxx_map_remove_key()

jdurgin replied:
apparently rocksdb_enable_rmrange is still experimental, though there may be more work on it coming soon (a couple folks attended a rocksdb meetup yesterday and asked the developers about it). So it sounds like we should keep it off for now.

Reworking this to list and then remove_range() like you suggest sounds good.

jdurgin commented on src/include/rados/librados.h:
```diff
@@ -3103,6 +3103,19 @@ CEPH_RADOS_API void rados_write_op_omap_rm_keys2(rados_write_op_t write_op,
+ * @param key_end the upper bound of the key range to remove
+ */
+CEPH_RADOS_API void rados_write_op_omap_rm_range(rados_write_op_t write_op,
+                                                 const char *key_begin,
```
to allow embedded NULs add size_t params for these keys as well - that's what the *omap2 functions do

neha-ojha commented:
@cbodley any update on this?

mattbenjamin commented:
@neha-ojha I forget the end state on this one, think we still care about it

jdurgin commented:
With this latest rocksdb update, rocksdb_enable_rmrange is stable: https://github.com/ceph/ceph/pull/29026

cbodley commented:
working through the rebase now

cbodley commented:
@liewegas @jdurgin @neha-ojha with the recent rocksdb update in https://github.com/ceph/ceph/pull/29026 and default support for rocksdb_enable_rmrange in https://github.com/ceph/ceph/pull/29323, i'd like to revisit this requirement to list keys in order to limit the number of trimmed entries per op

one concern with doing this is the possibility that the osd configuration overrides rocksdb_enable_rmrange=false. are there any objections to removing this rocksdb_enable_rmrange option entirely, so that cls and omap object deletion can rely on performant ranged deletes?

the current limitation of 1000 entries per osd request means that rgw has to carefully throttle its multisite log trimming to avoid noticeable impact on clients. in larger deployments, this makes it extremely difficult to keep up with log growth, especially when there's already a large backlog of entries

liewegas commented:
Yes, I'm 100% on board with removing the option entirely!

Unfortunately we probably can't removing the throttling logic for a while, since radosgw maybe running against older OSDs?

cbodley commented:
thanks @liewegas. agreed that backward compatability will be an issue. the semantics of these trim calls is that the client keeps resending the request until the osd returns ENODATA to indicate that the entire range was trimmed. these new calls don't have that semantic, so i think they should be named differently like cls_log_trim2

so a client can attempt to call cls_log_trim2 and, if not supported, fall back to resending cls_log_trim requests with throttling logic

in cls_log_trim, we still don't need to impose a limit on the number of keys trimmed, but we do need a way to tell whether the range was empty in order to return ENODATA. maybe there's a way for the objectstore to tell us that on OP_OMAP_RMKEYRANGE? if not, we can list keys in the range (with max=1)

liewegas commented:
> thanks @liewegas. agreed that backward compatability will be an issue. the semantics of these trim calls is that the client keeps resending the request until the osd returns ENODATA to indicate that the entire range was trimmed. these new calls don't have that semantic, so i think they should be named differently like cls_log_trim2
> 
> so a client can attempt to call cls_log_trim2 and, if not supported, fall back to resending cls_log_trim requests with throttling logic
> 
> in cls_log_trim, we still don't need to impose a limit on the number of keys trimmed, but we do need a way to tell whether the range was empty in order to return ENODATA. maybe there's a way for the objectstore to tell us that on OP_OMAP_RMKEYRANGE? if not, we can list keys in the range (with max=1)

I think listing with max=1 is the only way.  Sounds good!

cbodley commented:
i updated cls_log_trim() and cls_rgw_bilog_trim() to list the range with max=1 in order to detect the ENODATA case. the test cases are passing again, both against the old and new implementations

i decided not to introduce the new cls_log_trim2() interfaces for now, because we get most of the wins from rmrange itself; the client just has to send one extra osd op to see ENODATA, rather than the 1 + N/1000 requests needed before

i also added a final commit to remove the config options around rmrange - i'm happy to split that into another pr if necessary

cbodley commented:
this pass an rgw suite in http://pulpito.ceph.com/cbodley-2019-08-07_11:50:06-rgw-wip-19975-distro-basic-smithi/

should it run through rados?

cbodley commented:
results from a [very naive benchmark](https://gist.github.com/cbodley/183f9819d33afdc06fe36359ceab74f9) run against a debug-built vstart cluster with a single osd on nvme:
```
10k created entries in 0.727089
10k trimmed entries in 0.001970
100k created entries in 9.462839
100k trimmed entries in 0.002188
1m created entries in 100.594593
1m trimmed entries in 0.002140
10m created entries in 803.385073
10m trimmed entries in 0.002612
```

cbodley commented:
thanks @liewegas and @tchaikov - this passed the rgw suite, does it need to run through rados? packages are available at https://shaman.ceph.com/builds/ceph/wip-19975/ if you're willing to schedule it for me

tchaikov commented:
@cbodley i think Sage has included this PR in his batch =)

cbodley commented:
thanks for the help with review and testing!

vumrao commented:
nautilus backport #30949.
